README.md
# foundry-vtt-breadcrumbs

![Foundry VTT Breadcrumbs](artwork/module/breadcrumbs-cover.png)

## Summary
Foundry VTT Breadcrumbs adds configurable tile-based trails to the tokens of any actor that the GM desires. Breadcrumb images can be unique to scenes, actors, or individual tokens. Trails can be configured for length, style, and other properties both globally and per-scene.

# Configuration Options
## <a name="world-options"></a>World Options
`Breadcrumbs Enabled` Enabled Breadcrumb usage across any configured scenes.

`Default Image` The image used for crumbs unless overridden by scene or actor settings.

`Default Tint` The tint image used for crumbs unless overridden by scene or actor settings.

`Default Scale` The scale used for crumbs unless overridden by scene or actor settings.

`Max Trail Length` The maximum number of crumbs in a given trail. Can be overriden at the scene level.

`Hidden` Breadcrumbs trails are only visible to the GM role.

## <a name="scene-options"></a>Scene Options
#### To set these options, use the included `breadcrumbsSceneConfig` macro to apply settings to the currently-viewed scene.

`Enable Breadcrumbs` Enables the use of Breadcrumbs on the scene.

`Override Actors` Prefer the scene's default Breadcrumbs settings over any available actor settings. Note that this does _not_ override individual token settings, which may be set explicitly. See 

`Breadcrumbs Image` The default image for crumbs in the scene.

`Tint` The default tint for crumbs in the scene.

`Scale` The default scale for crumbs in the scene.

`Trail Length` A scene-specific override for the maxmimum trail length specified in [World Options](#world-options).

## <a name="actor-options"></a>Actor Options
#### To set these options, use the included `breadcrumbsActorConfig` macro with a token or multiple tokens selected.

`Enable Breadcrumbs` Enable Breadcrumbs for this actor or token.

`Apply to Actor` If left unchecked, the configuration options will be set on the token or tokens directly, bypassing any scene-specific configurations. If checked, the settings will propogate to the actor, becoming the default for all of that actor's newly-placed tokens and being subject to the `Override Actors` setting in [Scene Options](#scene-options).

`Breadcrumbs Image` The token or actor's default Breadcrumbs image.

`Image Tint` The token or actor's default Breadcrumbs tint.

`Image Scale` The token or actor's default Breadcrumbs scale. If this is not specified (i.e. '1') crumbs will scale along with the parent token's scale.

`Alternate Sides` Mirror every-other Breadcrumb created. This allows for a 'right-foot, left-foot' pattern where each Breadcrumb is a single step in the token's 'stride.' This is especially effective for giant footprints etc where each grid space moved through represents only a portion of the creature's overall token area.

`Hidden` An actor-specific override for the Hidden setting in [World Options](#world-options).

# Media
![Hex Crawls](artwork/module/breadcrumbs-hex.gif)

CHANGELOG.md
# foundry-vtt-breadcrumbs

## 1.0.5
- Added hidden (invisible) trail option

## 1.0.4
- Lots of code improvements
- Crumbs are now properly-aligned behind tokens regardless of scale
- Crumbs now scale with token size changes
- Support for alternating crumbs
- createBreadcrumb hook added
- README updates

## 1.0.1
- integrated dependency on color settings module for tint control
- completed settings for tint control
- reworked settings precedence for convenience and flexibility

## 1.0.0a
- initial alpha release of local code snippets

scripts/breadcrumbs.js
Hooks.on("createToken", function(tokenDocument, options, userId) {
    if((tokenDocument.actor.flags?.breadcrumbs?.enabled == true) || (tokenDocument.parent.flags?.breadcrumbs?.enabled == true)) {
          console.debug("Configuring new Breadcrumbs token")

        tokenDocument.update({
            flags: {
                breadcrumbs: {
                    enabled: true,
                    style: {
                      src: tokenDocument.parent.flags?.breadcrumbs?.override_actors === true 
                           ? tokenDocument.parent.flags?.breadcrumbs?.style?.src 
                           : (tokenDocument.actor.flags?.breadcrumbs?.style?.src || game.settings.get("breadcrumbs", "breadcrumbs-default-image")),
                      scale: tokenDocument.parent.flags?.breadcrumbs?.override_actors === true 
                             ? tokenDocument.parent.flags?.breadcrumbs?.style?.scale 
                             : (tokenDocument.actor.flags?.breadcrumbs?.style?.scale || game.settings.get("breadcrumbs", "breadcrumbs-default-scale")),
                      tint: tokenDocument.parent.flags?.breadcrumbs?.override_actors === true 
                            ? tokenDocument.parent.flags?.breadcrumbs?.style?.tint 
                            : (tokenDocument.actor.flags?.breadcrumbs?.style?.tint || game.settings.get("breadcrumbs", "breadcrumbs-default-tint")),
                      alternating: tokenDocument.parent.flags?.breadcrumbs?.override_actors === true 
                            ? tokenDocument.parent.flags?.breadcrumbs?.style?.alternating 
                            : (tokenDocument.actor.flags?.breadcrumbs?.style?.alternating || false),    
                    },
                    trail: {
                        id: tokenDocument.parent.id + "-" + tokenDocument.id,
                        timestamp: Date.now(),
                    },
                    position: {
                        last_x: tokenDocument.x, 
                        last_y: tokenDocument.y
                    }
                }
            }
        });
        tokenDocument.parent.update({
            flags: {
                breadcrumbs: {
                    trails: {
                        [tokenDocument.parent.id + "-" + tokenDocument.id]: {
                            totalCrumbs: 1
                        }
                    }
                }
            }
        });
    };
    Hooks.off("createToken");
});

function getRotationAngle(oldX, oldY, newX, newY) {
  let dx = newX - oldX;
  let dy = newY - oldY;

  // Figure out which direction the token moved in to orient the tile
  if (dx === 0 && dy > 0) {
    return 0;  // South
  } else if (dx > 0 && dy < 0) {
    return 225;  // North-East
  } else if (dx < 0 && dy > 0) {
    return 45;  // South-West
  } else if (dx > 0 && dy === 0) {
    return 270;  // West
  } else if (dx < 0 && dy < 0) {
    return 135;  // North-West
  } else if (dx === 0 && dy < 0) {
    return 180;  // North
  } else if (dx < 0 && dy === 0) {
    return 90;  // East
  } else if (dx > 0 && dy > 0) {
    return 315;  // South-East
  } else {
    return null;
  }
};

Hooks.on("updateToken", async function(tokenDocument, updateData, _, _) {
    let movementDirection = undefined
    const hasBreadcrumbsEnabled = tokenDocument.parent.flags?.breadcrumbs?.enabled == true && (
        tokenDocument.flags?.breadcrumbs?.enabled == true || tokenDocument.actor.flags?.breadcrumbs?.enabled == true);
    const hasPositionUpdate = updateData.y || updateData.x;
    
    if (hasBreadcrumbsEnabled && hasPositionUpdate) {
        // If the token has not moved at least half of its size, do not produce a Crumb
        if ((Math.abs(tokenDocument.x - tokenDocument.flags.breadcrumbs.position.last_x) < (tokenDocument.width * tokenDocument.parent.grid.size / 2)) && 
            (Math.abs(tokenDocument.y - tokenDocument.flags.breadcrumbs.position.last_y) < (tokenDocument.height * tokenDocument.parent.grid.size / 2))) 
            {
                return;
            } else {
                movementDirection = getRotationAngle(
                    tokenDocument.flags.breadcrumbs.position.last_x, 
                    tokenDocument.flags.breadcrumbs.position.last_y, 
                    tokenDocument.x, tokenDocument.y);
                tokenDocument.update({
                    flags: {
                        breadcrumbs: {
                            position: {
                                last_x: tokenDocument.x, 
                                last_y: tokenDocument.y
                            }
                        }
                    }
                });
            };
    } else { return; };

    function getMergedBreadcrumbsSettings(tokenDocument) {
      // Default settings
      const defaultSettings = {
          src: null,
          scale: undefined,
          tint: undefined,
      };
  
      const actorSettings = tokenDocument.actor.flags?.breadcrumbs?.style || {};
      const tokenSettings = tokenDocument.flags?.breadcrumbs?.style || {};
      const sceneSettings = tokenDocument.parent.flags?.breadcrumbs?.style || {};
  
      let mergedSettings = { ...defaultSettings };
  
      // If token settings are defined, they take precedence
      if (Object.keys(tokenSettings).length > 0) {
          mergedSettings = { ...mergedSettings, ...tokenSettings };
      } 
      // If token settings aren't defined, check for `override_actors`
      else if (tokenDocument.parent.flags?.breadcrumbs?.override_actors) {
          mergedSettings = { ...mergedSettings, ...sceneSettings };
      } 
      // If token settings aren't defined and `override_actors` isn't enabled, fall back to actor settings
      else {
          mergedSettings = { ...mergedSettings, ...actorSettings };
      }
  
      return mergedSettings;
    };  
    const actorSettings = getMergedBreadcrumbsSettings(tokenDocument);

    let maxTrailLength = tokenDocument.parent.flags.breadcrumbs?.trails?.length?.max || game.settings.get("breadcrumbs", "breadcrumbs-default-trail-length");
    let trailCrumbCount = tokenDocument.parent.flags?.breadcrumbs?.trails?.[tokenDocument.parent.id + "-" + tokenDocument.id].totalCrumbs
    const isAlternate = (trailCrumbCount + 1) % 2 !== 0 && tokenDocument.flags.breadcrumbs.style.alternating === true;

    breadcrumbsTileDefinition = {
        flags: {
            breadcrumbs: {
                trail: {
                    id: tokenDocument.parent.id + "-" + tokenDocument.id,
                    timestamp: Date.now()
                },
            }
        },
        texture: {
            src: actorSettings.src || game.settings.get("breadcrumbs", "breadcrumbs-default-image"),
            tint: actorSettings.tint || game.settings.get("breadcrumbs", "breadcrumbs-default-tint").substring(0, 7),
            scaleX: isAlternate ? -(actorSettings.scale || game.settings.get("breadcrumbs", "breadcrumbs-default-scale")) * tokenDocument.width: (actorSettings.scale || game.settings.get("breadcrumbs", "breadcrumbs-default-scale")) * tokenDocument.width,
            scaleY: actorSettings.scale * tokenDocument.height || game.settings.get("breadcrumbs", "breadcrumbs-default-scale") * tokenDocument.height,
            rotation: 0
        },
        x: tokenDocument.x + (tokenDocument.width * tokenDocument.parent.grid.size) / 2 - tokenDocument.parent.grid.size / 2,
        y: tokenDocument.y + (tokenDocument.height * tokenDocument.parent.grid.size) / 2 - tokenDocument.parent.grid.size / 2,
        height: tokenDocument.parent.grid.size,
        width: tokenDocument.parent.grid.size,
        hidden: actorSettings.hidden || game.settings.get("breadcrumbs", "breadcrumbs-hidden") || false,
        rotation: movementDirection
    };

    await tokenDocument.parent.createEmbeddedDocuments("Tile", [breadcrumbsTileDefinition]);

    tokenDocument.parent.update({
        flags: {
            breadcrumbs: {
                trails: {
                    [tokenDocument.parent.id + "-" + tokenDocument.id]: {
                        totalCrumbs: trailCrumbCount + 1
                    }
                }
            }
        }
    });
    
    Hooks.call('createBreadcrumb');

    let existingBreadcrumbs = tokenDocument.parent.tiles.filter(
        tile => tile.flags?.breadcrumbs?.trail?.id == tokenDocument.parent.id + "-" + tokenDocument.id);

    while (existingBreadcrumbs.length > maxTrailLength) {
        existingBreadcrumbs.sort((a, b) => a.flags.breadcrumbs.trail.timestamp - b.flags.breadcrumbs.trail.timestamp);
        let oldestTile = existingBreadcrumbs.shift();  // Removes the first (oldest) tile from the array
        oldestTile.delete();
    };

    Hooks.off("updateToken");
});

macros/breadcrumbsActorConfig.js
const currentScene = game.scenes.current

const loadoutsTileDialog = new Dialog({
    title: "Breadcrumbs", 
    content:`
    <script>
        async function fileBrowser() {
            const path = '` + game.settings.get("breadcrumbs", "breadcrumbs-default-image") + `';
            new FilePicker({
                type: 'image',
                current: path,
                callback: imagePath => {
                    document.getElementById("breadcrumbsImagePreview").src = imagePath;
                },
            }).browse();
    }
    </script>
    <script type="text/javascript">
        function limitTrailLength(obj){
            if (obj.value.length > 2){
                obj.value = obj.value.slice(0,2); 
            }
        }
    </script>
    <form class="form-horizontal">
    <fieldset>
    
    <!-- Form Name -->
    <legend>Actor & Token Controls</legend>

    <!-- Boolean input-->
    <div class="form-group">
      <label class="col-md-4 control-label" for="enableBreadcrumbs">Enable Breadcrumbs</label>  
      <div class="col-md-4">
      <input id="enableBreadcrumbs" name="enableBreadcrumbs" type="checkbox" checked></input>
      </div>
    </div>
    
    <div class="form-group">
      <label class="col-md-4 control-label" for="applyToActor">Apply to Actor</label>  
      <div class="col-md-4">
      <input id="applyToActor" name="applyToActor" type="checkbox"></input>
      </div>
    </div>

    </fieldset>
    </form>

    <form class="form-horizontal">
    <fieldset>
    <legend>Image Properties</legend>
    
    <div class="form-group">
        <label>Breadcrumbs Image</label>
        <button onclick="fileBrowser()">
            <img id="breadcrumbsImagePreview" src="` + game.settings.get("breadcrumbs", "breadcrumbs-default-image") + `" style="width: 50px; height: 50px;">
        </button>
    </div>

    <div class="form-group">
        <label>Image Tint</label>
        <input type="text" name="imageTint" is="colorpicker-input" data-responsive-color>
    </div>

    <!-- Range input-->
    <div class="form-group">
      <label class="col-md-4 control-label" for="imageScale">Image Scale</label>  
      <div class="col-md-4">
        <input id="imageScale" name="imageScale" type="range" min="0.1" max="1" step="0.1" defaultValue="1" value="1" oninput="document.getElementById('rangeValLabel').innerHTML = this.value;"></input>
        <span class="help-block">Scale: </span>
      <em id="rangeValLabel" style="font-style: normal;">1</em>
      </div>
    </div>

    </fieldset>
    </form>

    <form class="form-horizontal">
    <fieldset>
    <legend>Trail Properties</legend>

    <div class="form-group">
      <label class="col-md-4 control-label" for="alternateSidess">Alternate Sides</label>
      <div class="col-md-4">
      <input id="alternateSides" name="alternateSides" type="checkbox"></input>
      </div>
    </div>

    <div class="form-group">
      <label class="col-md-4 control-label" for="isHidden">Hidden</label>
      <div class="col-md-4">
      <input id="isHidden" name="isHidden" type="checkbox"></input>
      </div>
    </div>

    </fieldset>
    </form>
    `,
    buttons: {
    cancel: {
        icon: "<i class='fas fa-check'></i>",
        label: `Cancel`,
        callback: function(){ return; }  
        },
    apply: {
        icon: "<i class='fas fa-check'></i>",
        label: `Apply Changes` ,
        callback: html => {setupBreadcrumbsActors(
            html.find('[name="enableBreadcrumbs"]').val(),
            html.find('[name="applyToActor"]').val(),
            html.find("#breadcrumbsImagePreview").attr("src"),
            html.find('[name="imageTint"]').val(),
            html.find('[name="imageScale"]').val(),
            html.find('[name="alternateSides"]').val(),
            html.find('[name="isHidden"]').val(),
        )}   
        }
      },
    default: 'apply',
}).render(true);

async function setupBreadcrumbsActors(enableBreadcrumbs, applyToActor, breadcrumbsImage, imageTint, imageScale, alternateSides, isHidden) {
    const enableBreadcrumbsCheckbox = document.getElementById('enableBreadcrumbs');
    const setForSceneCheckbox = document.getElementById('setForScene');
    const alternateSidesCheckbox = document.getElementById('alternateSides');
    const isHiddenCheckbox = document.getElementById('alternateSides');

    if(applyToActor.checked) {
        canvas.tokens.controlled.forEach(token => token.document.actor.update({
            flags: {
                breadcrumbs: {
                    enabled: enableBreadcrumbsCheckbox.checked,
                    style: {
                        src: breadcrumbsImage,
                        scale: imageScale,
                        tint: imageTint.substring(0, 7),
                        alternating: alternateSidesCheckbox.checked,
                        hidden: isHiddenCheckbox.checked
                    }
                }
            }
        }));
    } else {
        canvas.tokens.controlled.forEach(token => token.document.update({
            flags: {
                breadcrumbs: {
                    enabled: enableBreadcrumbsCheckbox.checked,
                    style: {
                        src: breadcrumbsImage,
                        scale: imageScale,
                        tint: imageTint.substring(0, 7),
                        alternating: alternateSidesCheckbox.checked,
                        hidden: isHiddenCheckbox.checked
                    }
                }
            }
        }));
    };
};

macros/breadcrumbsSceneConfig.js
const currentScene = game.scenes.current

const loadoutsTileDialog = new Dialog({
    title: "Breadcrumbs Scene Configuration", 
    content:`
    <script>
        async function fileBrowser() {
            const path = '` + game.settings.get("breadcrumbs", "breadcrumbs-default-image") + `';
            new FilePicker({
                type: 'image',
                current: path,
                callback: imagePath => {
                    document.getElementById("breadcrumbsImagePreview").src = imagePath;
                },
            }).browse();
    }
    </script>
    <script type="text/javascript">
        function limitTrailLength(obj){
            if (obj.value.length > 2){
                obj.value = obj.value.slice(0,2); 
            }
        }
    </script>
    <form class="form-horizontal">
    <fieldset>
    
    <!-- Form Name -->
    <legend>Scene Settings</legend>

    <!-- Boolean input-->
    <div class="form-group">
      <label class="col-md-4 control-label" for="enableBreadcrumbs">Enable Breadcrumbs</label>  
      <div class="col-md-4">
      <input id="enableBreadcrumbs" name="enableBreadcrumbs" type="checkbox" checked></input>
      </div>
    </div>

    <div class="form-group">
      <label class="col-md-4 control-label" for="overrideActors">Override Actors</label>  
      <div class="col-md-4">
      <input id="overrideActors" name="overrideActors" type="checkbox"></input>
      </div>
    </div>

    </fieldset>
    </form>
    
    <form class="form-horizontal">
    <fieldset>
    <legend>Image Properties</legend>

    <div class="form-group">
        <label>Breadcrumbs Image</label>
        <button onclick="fileBrowser()">
            <img id="breadcrumbsImagePreview" src="` + game.settings.get("breadcrumbs", "breadcrumbs-default-image") + `" style="width: 50px; height: 50px;">
        </button>
    </div>

    <div class="form-group">
        <label>Image Tint</label>
        <input type="text" name="imageTint" is="colorpicker-input" data-responsive-color>
    </div>

    <!-- Range input-->
    <div class="form-group">
      <label class="col-md-4 control-label" for="breadcrumbsScale">Breadcrumbs Scale</label>  
      <div class="col-md-4">
      <input id="breadcrumbsScale" name="breadcrumbsScale" type="range" min="0.1" max="1" step="0.1" defaultValue="1" value="1" oninput="document.getElementById('rangeValLabel').innerHTML = this.value;"></input>
      <span class="help-block">Scale: </span>
      <em id="rangeValLabel" style="font-style: normal;">1</em>
      </div>
    </div>

    </fieldset>
    </form>

    <form class="form-horizontal">
    <fieldset>
    <legend>Trail Properties</legend>

    <div class="form-group">
      <label for="trailLength">Max Trail Length</label>
      <input type="number" id="trailLength" name="trailLength" value="` + game.settings.get("breadcrumbs", "breadcrumbs-default-trail-length") + `" min="0" max="99" oninput="limitTrailLength(this)">
    </div>

    <div class="form-group">
      <label class="col-md-4 control-label" for="alternateSidess">Alternate Sides</label>
      <div class="col-md-4">
        <input id="alternateSides" name="alternateSides" type="checkbox"></input>
      </div>
    </div>

    </fieldset>
    </form>
    `,
      buttons: {
        cancel: {
            icon: "<i class='fas fa-check'></i>",
            label: `Cancel`,
            callback: function(){ return; }  
            },
        apply: {
            icon: "<i class='fas fa-check'></i>",
            label: `Apply Changes` ,
            callback: html => {setupBreadcrumbsScene(
                html.find('[name="enableBreadcrumbs"]').val(),
                html.find('[name="overrideActors"]').val(),
                html.find("#breadcrumbsImagePreview").attr("src"),
                html.find('[name="imageTint"]').val(),
                html.find('[name="breadcrumbsScale"]').val(),
                html.find('[name="trailLength"]').val(),
                html.find('[name="alternateSides"]').val()
            )}   
            }
      },
      default: 'apply',
}).render(true);

async function setupBreadcrumbsScene(enableBreadcrumbs, overrideActors, breadcrumbsImage, imageTint, breadcrumbsScale, trailLength, alternateSides){
    const enableBreadcrumbsCheckbox = document.getElementById('enableBreadcrumbs');
    const overrideActorsCheckbox = document.getElementById('overrideActors');
    const alternateSidesCheckbox = document.getElementById('alternateSides');

    currentScene.update({
        "flags.breadcrumbs": {
            enabled: enableBreadcrumbsCheckbox.checked,
            override_actors: overrideActorsCheckbox.checked,
            style: {
                src: breadcrumbsImage,
                scale: breadcrumbsScale,
                tint: imageTint.substring(0, 7),
                alternateSides: alternateSidesCheckbox.checked
            },
            trails: {
                length: {
                    max: Number(trailLength)
                }

            }
        }
    });
};

macros/configureBreadcrumbsScene.js
const currentScene = game.scenes.current

const loadoutsTileDialog = new Dialog({
    title: "Breadcrumbs", 
    content:`
    <script>
        async function fileBrowser() {
            const path = '` + game.settings.get("breadcrumbs", "breadcrumbs-default-image") + `';
            new FilePicker({
                type: 'image',
                current: path,
                callback: imagePath => {
                    document.getElementById("breadcrumbsImagePreview").src = imagePath;
                },
            }).browse();
    }
    </script>
    <script type="text/javascript">
        function limitTrailLength(obj){
            if (obj.value.length > 2){
                obj.value = obj.value.slice(0,2); 
            }
        }
    </script>
    <form class="form-horizontal">
    <fieldset>
    
    <!-- Form Name -->
    <legend>Scene Configuration</legend>

    <!-- Boolean input-->
    <div class="form-group">
      <label class="col-md-4 control-label" for="enableBreadcrumbs">Enable Breadcrumbs</label>  
      <div class="col-md-4">
      <input id="enableBreadcrumbs" name="enableBreadcrumbs" type="checkbox" checked></input>
      </div>
    </div>
    
    <div class="form-group">
      <label class="col-md-4 control-label" for="overrideActors">Override Actors</label>  
      <div class="col-md-4">
      <input id="overrideActors" name="overrideActors" type="checkbox"></input>
      </div>
    </div>

    <div class="form-group">
        <label>Breadcrumbs Image</label>
        <button onclick="fileBrowser()">
            <img id="breadcrumbsImagePreview" src="` + game.settings.get("breadcrumbs", "breadcrumbs-default-image") + `" style="width: 50px; height: 50px;">
        </button>
    </div>

    <!-- Range input-->
    <div class="form-group">
      <label class="col-md-4 control-label" for="breadcrumbsScale">Breadcrumbs Scale</label>  
      <div class="col-md-4">
      <input id="breadcrumbsScale" name="breadcrumbsScale" type="range" min="0.1" max="1" step="0.1" defaultValue="1" value="1" oninput="document.getElementById('rangeValLabel').innerHTML = this.value;"></input>
      <span class="help-block">Scale: </span>
      <em id="rangeValLabel" style="font-style: normal;">1</em>
      </div>
    </div>

    <div class="form-group">
      <label for="trailLength">Max Trail Length</label>
      <input type="number" id="trailLength" name="trailLength" value="` + game.settings.get("breadcrumbs", "breadcrumbs-default-trail-length") + `" min="0" max="99" oninput="limitTrailLength(this)">
    </div>
    `,
      buttons: {
        cancel: {
            icon: "<i class='fas fa-check'></i>",
            label: `Cancel`,
            callback: function(){ return; }  
            },
        apply: {
            icon: "<i class='fas fa-check'></i>",
            label: `Apply Changes` ,
            callback: html => {setupBreadcrumbsScene(
                html.find('[name="enableBreadcrumbs"]').val(),
                html.find('[name="overrideActors"]').val(),
                html.find("#breadcrumbsImagePreview").attr("src"),
                html.find('[name="breadcrumbsScale"]').val(),
                html.find('[name="trailLength"]').val()
            )}   
            }
      },
      default: 'apply',
}).render(true);

async function setupBreadcrumbsScene(enableBreadcrumbs, overrideActors, breadcrumbsImage, breadcrumbsScale, trailLength){
    const enableBreadcrumbsCheckbox = document.getElementById('enableBreadcrumbs');
    const overrideActorsCheckbox = document.getElementById('overrideActors');
    currentScene.update({
        "flags.breadcrumbs": {
            enabled: enableBreadcrumbsCheckbox.checked,
            override_actors: overrideActorsCheckbox.checked,
            actors: {
                default: {
                    src: breadcrumbsImage,
                    scale: breadcrumbsScale,
                    tint: null
                }
            },
            trails: {
                length: {
                    max: Number(trailLength)
                }

            }
        }
    });
};

macros/configureBreadcrumbsActors.js
const currentScene = game.scenes.current

const loadoutsTileDialog = new Dialog({
    title: "Breadcrumbs", 
    content:`
    <script>
        async function fileBrowser() {
            const path = '` + game.settings.get("breadcrumbs", "breadcrumbs-default-image") + `';
            new FilePicker({
                type: 'image',
                current: path,
                callback: imagePath => {
                    document.getElementById("breadcrumbsImagePreview").src = imagePath;
                },
            }).browse();
    }
    </script>
    <script type="text/javascript">
        function limitTrailLength(obj){
            if (obj.value.length > 2){
                obj.value = obj.value.slice(0,2); 
            }
        }
    </script>
    <form class="form-horizontal">
    <fieldset>
    
    <!-- Form Name -->
    <legend>Actor Configuration</legend>

    <!-- Boolean input-->
    <div class="form-group">
      <label class="col-md-4 control-label" for="enableBreadcrumbs">Enable Breadcrumbs</label>  
      <div class="col-md-4">
      <input id="enableBreadcrumbs" name="enableBreadcrumbs" type="checkbox" checked></input>
      </div>
    </div>
    
    <div class="form-group">
      <label class="col-md-4 control-label" for="setForScene">Limit to Scene?</label>  
      <div class="col-md-4">
      <input id="setForScene" name="setForScene" type="checkbox"></input>
      </div>
    </div>

    <div class="form-group">
        <label>Breadcrumbs Image</label>
        <button onclick="fileBrowser()">
            <img id="breadcrumbsImagePreview" src="` + game.settings.get("breadcrumbs", "breadcrumbs-default-image") + `" style="width: 50px; height: 50px;">
        </button>
    </div>

    <!-- Range input-->
    <div class="form-group">
      <label class="col-md-4 control-label" for="breadcrumbsScale">Breadcrumbs Scale</label>  
      <div class="col-md-4">
      <input id="breadcrumbsScale" name="breadcrumbsScale" type="range" min="0.1" max="1" step="0.1" defaultValue="1" value="1" oninput="document.getElementById('rangeValLabel').innerHTML = this.value;"></input>
      <span class="help-block">Scale: </span>
      <em id="rangeValLabel" style="font-style: normal;">1</em>
      </div>
    </div>
    `,
    buttons: {
    cancel: {
        icon: "<i class='fas fa-check'></i>",
        label: `Cancel`,
        callback: function(){ return; }  
        },
    apply: {
        icon: "<i class='fas fa-check'></i>",
        label: `Apply Changes` ,
        callback: html => {setupBreadcrumbsActors(
            html.find('[name="enableBreadcrumbs"]').val(),
            html.find('[name="setForScene"]').val(),
            html.find("#breadcrumbsImagePreview").attr("src"),
            html.find('[name="breadcrumbsScale"]').val()
        )}   
        }
      },
    default: 'apply',
}).render(true);

async function setupBreadcrumbsActors(enableBreadcrumbs, setForScene, breadcrumbsImage, breadcrumbsScale) {
    const enableBreadcrumbsCheckbox = document.getElementById('enableBreadcrumbs');
    const setForSceneCheckbox = document.getElementById('setForScene');

    if(!setForSceneCheckbox.checked) {
        canvas.tokens.controlled.forEach(token => token.document.actor.update({
            flags: {
                breadcrumbs: {
                    enabled: enableBreadcrumbsCheckbox.checked,
                    style: {
                        src: breadcrumbsImage,
                        scale: breadcrumbsScale,
                        tint: null
                    }
                }
            }
        }))
    } else {
        canvas.tokens.controlled.forEach(token => token.document.update({
            flags: {
                breadcrumbs: {
                    enabled: enableBreadcrumbsCheckbox.checked,
                }
            }
        }))
        canvas.tokens.controlled.forEach(token => currentScene.update({
            flags: {
                breadcrumbs: {
                    actors: {
                        [token.document.actor.id]: {
                            src: breadcrumbsImage,
                            scale: breadcrumbsScale,
                            tint: null
                        }
                    }
                }
            }
        }));
    };
};

scripts/config.js
Hooks.once("init", function () {

    game.settings.register("breadcrumbs", "breadcrumbs-enabled", {
        name: "Enable Breadcrumbs",
        hint: "Enable the use of Breadcrumbs",
        scope: "world",
        config: true,
        default: true,
        type: Boolean
    });

    game.settings.register("breadcrumbs", "breadcrumbs-default-image", {
        name: "Default Image",
        hint: "Per-scene images can be configured using the scene configuration macro",
        scope: "world",
        config: true,
        filePicker: true,
        default: 'modules/breadcrumbs/artwork/shoeprints.webp',
        type: String
    });

    new window.Ardittristan.ColorSetting("breadcrumbs", "breadcrumbs-default-tint", {
        name: "Default Tint",
        hint: "A default tint for crumbs",
        label: "Color Picker",
        restricted: false,
        defaultColor: null,
        scope: "world",
        //onChange: (value) => {}
    });

    game.settings.register("breadcrumbs", "breadcrumbs-default-scale", {
        name: "Default Image Scale",
        hint: "Scale of Breadcrumbs images",
        scope: "world",
        config: true,
        default: 1,
        type: Number,
        range: {
            min: 0.1,
            max: 1,
            step: .1
        }
    });

    game.settings.register("breadcrumbs", "breadcrumbs-default-trail-length", {
        name: "Max Trail Length",
        hint: "The maximum number of Breadcrumbs allowed in a trail",
        scope: "world",
        config: true,
        default: 50,
        type: Number,
        range: {
            min: 5,
            max: 99,
            step: 1
        }
    });

    game.settings.register("breadcrumbs", "breadcrumbs-hidden", {
        name: "Hidden",
        hint: "Trails are only visible to the GM role",
        scope: "world",
        config: true,
        default: false,
        type: Boolean,
    });

});
